server/sonar-server/src/main/java/org/sonar/server/issue/IssueQueryFactory.java

```diff
@@ -364,24 +363,13 @@ private static void addDirectories(IssueQuery.Builder builder, List<ComponentDto
     return componentDtos;
   }
 
-  @VisibleForTesting
-  static Collection<RuleKey> toRules(@Nullable Object o) {
-    Collection<RuleKey> result = null;
-    if (o != null) {
-      if (o instanceof List) {
-        // assume that it contains only strings
-        result = stringsToRules((List<String>) o);
-      } else if (o instanceof String) {
-        result = stringsToRules(newArrayList(Splitter.on(',').omitEmptyStrings().split((String) o)));
-      }
-    }
-    return result;
-  }
-
   @CheckForNull
-  private static Collection<RuleKey> stringsToRules(@Nullable Collection<String> rules) {
+  private Collection<Integer> ruleKeysToRuleId(DbSession dbSession, @Nullable Collection<String> rules) {
     if (rules != null) {
-      return Collections2.transform(rules, RuleKey::parse);
+      return dbClient.ruleDao().selectDefinitionByKeys(dbSession, transform(rules, RuleKey::parse))
```

Comment:
this can and must be optimized.
Those rules are loaded twice: in IssueQueryBuilder and SearchResponseLoader.

see https://github.com/SonarSource/sonarqube/pull/2994